File: r/R/response.R
# The response shapes the package takes, in the forms the R ecosystem
# already uses, and the outcome family each selects: `glm`'s rule, the
# declared family authoritative and the response checked against it, with
# "not declared" representable.

#' Parse a response into the shape the core takes
#'
#' A `survival::Surv` object of type `"right"` selects the AFT family and
#' one of type `"interval"` (the `"interval2"` encoding) the
#' interval-censored family; an ordered factor selects the ordinal family,
#' a two-level factor the probit family, and a numeric vector the Gaussian
#' family.
#'
#' @param y The response as the caller gave it.
#' @param call The calling environment to report.
#' @return An object of class `"thiessen_response"`: a list of `kind`, the
#'   core entry point the response reaches (`"plain"`, `"aft"` or
#'   `"interval_censored"`); `shape`, the shape of the response;
#'   `family`, the outcome family the shape selects; `n`, the number of
#'   observations; the numeric vector or the censoring columns the core
#'   takes; and the levels of a factor response.
#' @noRd
as_response <- function(y, call = rlang::caller_env()) {
  if (inherits(y, "Surv")) {
    return(surv_response(y, call = call))
  }
  if (is.factor(y)) {
    return(factor_response(y, call = call))
  }
  if (!is.numeric(y) || !is.null(dim(y))) {
    thiessen_abort(
      paste0(
        "`y` must be a numeric vector, a factor or a `survival::Surv` ",
        "object."
      ),
      call = call
    )
  }
  if (anyNA(y)) {
    thiessen_abort("`y` must not contain missing values.", call = call)
  }
  new_response("plain", "numeric", y = as.double(y))
}

#' A factor response: ordered as ordinal codes, two levels as labels
#'
#' The codes are 0 to K - 1 with the first level as 0, as `glm` treats a
#' factor response and `MASS::polr` an ordered one.
#'
#' @param y A factor.
#' @param call The calling environment to report.
#' @return An object of class `"thiessen_response"`.
#' @noRd
factor_response <- function(y, call) {
  if (anyNA(y)) {
    thiessen_abort("`y` must not contain missing values.", call = call)
  }
  codes <- as.double(as.integer(y) - 1L)
  if (is.ordered(y)) {
    return(new_response("plain", "ordered", y = codes, levels = levels(y)))
  }
  if (nlevels(y) != 2L) {
    thiessen_abort(
      "A factor response must have two levels, or be an ordered factor.",
      call = call
    )
  }
  new_response("plain", "binary", y = codes, levels = levels(y))
}

#' A `Surv` response as the censoring columns the core takes
#'
#' A `Surv` is a numeric matrix with a `type` attribute. Type `"right"`
#' holds `time` and `status` (1 an event, 0 right-censored). Type
#' `"interval"`, which `Surv(lower, upper, type = "interval2")` also
#' produces, holds `time1`, `time2` and a `status` of 0 (right-censored
#' at `time1`), 1 (exact at `time1`), 2 (left-censored at `time1`) or 3
#' (between `time1` and `time2`); the bounds reach the core with an
#' infinite endpoint for one-sided censoring and an equal pair for an
#' exact value.
#'
#' @param y A `Surv` object.
#' @param call The calling environment to report.
#' @return An object of class `"thiessen_response"`.
#' @noRd
surv_response <- function(y, call) {
  type <- attr(y, "type")
  columns <- unclass(y)
  if (identical(type, "right")) {
    times <- columns[, "time"]
    status <- columns[, "status"]
    if (anyNA(times) || anyNA(status)) {
      thiessen_abort("`y` must not contain missing values.", call = call)
    }
    return(new_response(
      "aft", "right",
      times = as.double(times), events = status == 1
    ))
  }
  if (identical(type, "interval")) {
    time1 <- columns[, "time1"]
    time2 <- columns[, "time2"]
    status <- columns[, "status"]
    if (anyNA(time1) || anyNA(status) || anyNA(time2[status == 3])) {
      thiessen_abort("`y` must not contain missing values.", call = call)
    }
    lower <- ifelse(status == 2, -Inf, time1)
    upper <- ifelse(status == 3, time2, ifelse(status == 0, Inf, time1))
    return(new_response(
      "interval_censored", "interval",
      lower = as.double(lower), upper = as.double(upper)
    ))
  }
  thiessen_abort(
    sprintf(
      "A `Surv` response must be of type \"right\" or \"interval\", not \"%s\".",
      type
    ),
    call = call
  )
}

#' Construct a parsed response
#'
#' @param kind The core entry point: `"plain"`, `"aft"` or
#'   `"interval_censored"`.
#' @param shape The response shape: `"numeric"`, `"binary"`, `"ordered"`,
#'   `"right"` or `"interval"`.
#' @param y The numeric response of a plain kind.
#' @param levels The levels of a factor response.
#' @param times,events The columns of a right-censored response.
#' @param lower,upper The columns of an interval-censored response.
#' @return An object of class `"thiessen_response"`.
#' @noRd
new_response <- function(kind, shape, y = NULL, levels = NULL, times = NULL,
                         events = NULL, lower = NULL, upper = NULL) {
  structure(
    list(
      kind = kind,
      shape = shape,
      family = SHAPE_FAMILY[[shape]],
      n = NROW(if (kind == "plain") y else if (kind == "aft") times else lower),
      y = y,
      levels = levels,
      times = times,
      events = events,
      lower = lower,
      upper = upper
    ),
    class = "thiessen_response"
  )
}

# The outcome family each response shape selects when none is named.
SHAPE_FAMILY <- c(
  numeric = "gaussian",
  binary = "probit",
  ordered = "ordinal",
  right = "aft",
  interval = "interval_censored"
)

# The response shapes each outcome family accepts when named: the
# families over a real line take a numeric vector, the probit family the
# labels of a two-level factor or the numbers 0 and 1, and each other
# family the one shape that selects it.
FAMILY_SHAPES <- list(
  gaussian = "numeric",
  tobit = "numeric",
  student_t = "numeric",
  laplace = "numeric",
  probit = c("binary", "numeric"),
  ordinal = "ordered",
  aft = "right",
  interval_censored = "interval"
)

#' Describe a response shape in an error message
#'
#' @param shape A response shape.
#' @return A noun phrase.
#' @noRd
shape_label <- function(shape) {
  switch(
    shape,
    numeric = "a numeric vector",
    binary = "a two-level factor",
    ordered = "an ordered factor",
    right = "a `Surv` object of type \"right\"",
    interval = "a `Surv` object of type \"interval\""
  )
}

#' Refuse a named outcome family the response does not fit
#'
#' @param kind The core's name for the named family.
#' @param response An object of class `"thiessen_response"`.
#' @param call The calling environment to report.
#' @noRd
check_outcome_response <- function(kind, response,
                                   call = rlang::caller_env()) {
  if (response$shape %in% FAMILY_SHAPES[[kind]]) {
    return(invisible(NULL))
  }
  thiessen_abort(
    sprintf(
      "The response is %s, which selects the %s family, but `outcome` names the %s family.",
      shape_label(response$shape), response$family, kind
    ),
    call = call
  )
}

#' The control with its outcome family resolved against the response
#'
#' `NULL` takes the family the response selects; a named family is checked
#' against the response and never coerced. The ordinal family takes its
#' category count from the levels of the response where the constructor
#' left it unset.
#'
#' @param control An object of class `"thiessen_control"`.
#' @param response An object of class `"thiessen_response"`.
#' @param call The calling environment to report.
#' @return The control with a named outcome family.
#' @noRd
resolve_outcome <- function(control, response, call = rlang::caller_env()) {
  outcome <- control$outcome
  if (is.null(outcome)) {
    control$outcome <- family_of(response)
    return(control)
  }
  kind <- attr(outcome, "kind")
  check_outcome_response(kind, response, call = call)
  if (kind == "ordinal") {
    categories <- length(response$levels)
    if (is.null(outcome$categories)) {
      outcome$categories <- categories
    } else if (outcome$categories != categories) {
      thiessen_abort(
        sprintf(
          "`outcome` names %d categories but the response has %d levels.",
          outcome$categories, categories
        ),
        call = call
      )
    }
    control$outcome <- outcome
  }
  control
}

#' The outcome family a response selects, at its defaults
#'
#' @param response An object of class `"thiessen_response"`.
#' @return An object of class `"thiessen_outcome"`.
#' @noRd
family_of <- function(response) {
  switch(
    response$family,
    gaussian = gaussian_outcome(),
    probit = probit_outcome(),
    ordinal = ordinal_outcome(categories = length(response$levels)),
    aft = aft_outcome(),
    interval_censored = interval_censored_outcome()
  )
}

#' Construct the core's sampler over a parsed response
#'
#' @param config The configuration JSON.
#' @param design The numeric design.
#' @param response An object of class `"thiessen_response"`.
#' @param seed The resolved seed.
#' @param chain The chain index, from 0.
#' @return An external pointer to the sampler.
#' @noRd
new_sampler_handle <- function(config, design, response, seed, chain) {
  switch(
    response$kind,
    plain = core_sampler_new(config, design, response$y, seed, chain),
    aft = core_sampler_new_aft(
      config, design, response$times, response$events, seed, chain
    ),
    interval_censored = core_sampler_new_interval_censored(
      config, design, response$lower, response$upper, seed, chain
    )
  )
}

#' Replace a sampler's response with a parsed one
#'
#' @param handle An external pointer to the sampler.
#' @param response An object of class `"thiessen_response"`.
#' @noRd
set_sampler_response <- function(handle, response) {
  switch(
    response$kind,
    plain = core_sampler_set_response(handle, response$y),
    aft = core_sampler_set_aft_response(
      handle, response$times, response$events
    ),
    interval_censored = core_sampler_set_interval_censored_response(
      handle, response$lower, response$upper
    )
  )
}

#' The pointwise log-likelihood of a parsed response
#'
#' @param state An external pointer to the fitted state.
#' @param design The numeric design.
#' @param response An object of class `"thiessen_response"`.
#' @return A double matrix, one row per kept draw.
#' @noRd
log_lik_of <- function(state, design, response) {
  switch(
    response$kind,
    plain = core_log_lik(state, design, response$y),
    aft = core_log_lik_survival(
      state, design, response$times, response$events
    ),
    interval_censored = core_log_lik_interval_censored(
      state, design, response$lower, response$upper
    )
  )
}

File: python/python/thiessen/_response.py
"""The response shapes the package takes and the outcome family each selects.

The shapes are the ones the Python ecosystem already uses, and the rule is
``glm``'s: the response selects the family where none is named, a named
family is checked against the response, and nothing is coerced. A numeric
array selects the Gaussian family, a boolean array or a two-category
``Categorical`` the probit family, an ordered ``Categorical`` the ordinal
family, a structured survival array (the scikit-survival layout) the AFT
family, and a two-column array of bounds the interval-censored family.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

import numpy as np
import numpy.typing as npt

from . import _native
from ._arrays import _as_numeric
from ._params import Outcome
from .families import Aft, Gaussian, IntervalCensored, Ordinal, Probit

__all__ = [
    "Response",
    "_as_response",
    "_check_family",
    "_family_name",
    "_fit",
    "_log_likelihood",
    "_new_sampler",
    "_resolve_outcome",
    "_sampler_family",
    "_set_response",
]

Float = npt.NDArray[np.float64]
Bool = npt.NDArray[np.bool_]


def _empty() -> Float:
    return np.empty(0, dtype=np.float64)


def _no_flags() -> Bool:
    return np.empty(0, dtype=np.bool_)


@dataclass(frozen=True)
class Response:
    """A parsed response: the core entry point it reaches and its columns.

    Attributes
    ----------
    kind : str
        The core entry point: ``'plain'``, ``'aft'`` or
        ``'interval_censored'``.
    shape : str
        The response shape: ``'numeric'``, ``'binary'``, ``'ordered'``,
        ``'right'`` or ``'interval'``.
    y : numpy.ndarray
        The numeric response of a plain kind; the category codes of a
        ``Categorical``.
    times, events : numpy.ndarray
        The columns of a right-censored response.
    lower, upper : numpy.ndarray
        The columns of an interval-censored response.
    categories : tuple or None
        The categories of a ``Categorical`` response, in order.
    """

    kind: str
    shape: str
    y: Float = field(default_factory=_empty)
    times: Float = field(default_factory=_empty)
    events: Bool = field(default_factory=_no_flags)
    lower: Float = field(default_factory=_empty)
    upper: Float = field(default_factory=_empty)
    categories: tuple[Any, ...] | None = None

    @property
    def family(self) -> str:
        """str: The outcome family the shape selects."""
        return SHAPE_FAMILY[self.shape]

    @property
    def n(self) -> int:
        """int: The number of observations."""
        if self.kind == "aft":
            return len(self.times)
        if self.kind == "interval_censored":
            return len(self.lower)
        return len(self.y)


#: The outcome family each response shape selects when none is named.
SHAPE_FAMILY = {
    "numeric": "gaussian",
    "binary": "probit",
    "ordered": "ordinal",
    "right": "aft",
    "interval": "interval_censored",
}

#: The response shapes each outcome family accepts when named: the families
#: over a real line take a numeric array, the probit family a boolean array
#: or the numbers 0 and 1, the ordinal family an ordered `Categorical` or
#: integer codes, and each censored family the one shape that selects it.
FAMILY_SHAPES = {
    "gaussian": ("numeric",),
    "tobit": ("numeric",),
    "student_t": ("numeric",),
    "laplace": ("numeric",),
    "probit": ("binary", "numeric"),
    "ordinal": ("ordered", "numeric"),
    "aft": ("right",),
    "interval_censored": ("interval",),
}

SHAPE_LABEL = {
    "numeric": "a numeric array",
    "binary": "a boolean or two-category response",
    "ordered": "an ordered Categorical",
    "right": "a structured survival array",
    "interval": "a two-column array of bounds",
}


def _categorical(y: Any) -> Any:
    """Return the pandas categorical behind `y`, or `None`.

    A `Series` of category dtype answers through its ``cat`` accessor and
    a `Categorical` directly; both carry ``codes``, ``categories`` and
    ``ordered``.
    """
    if str(getattr(y, "dtype", "")) != "category":
        return None
    return getattr(y, "cat", y)


def _categorical_response(categorical: Any) -> Response:
    """Parse a pandas categorical: ordered as ordinal codes, two as labels.

    The codes are 0 to K - 1 in category order, the scikit-learn
    ``OrdinalEncoder`` convention.
    """
    codes = np.asarray(categorical.codes)
    if np.any(codes < 0):
        raise ValueError("y must not contain missing values")
    categories = tuple(categorical.categories)
    values = np.ascontiguousarray(codes, dtype=np.float64)
    if categorical.ordered:
        return Response("plain", "ordered", y=values, categories=categories)
    if len(categories) != 2:
        raise ValueError(
            "a Categorical response must have two categories, or be ordered; "
            f"got {len(categories)} unordered categories"
        )
    return Response("plain", "binary", y=values, categories=categories)


def _survival_response(y: Any) -> Response:
    """Parse a structured survival array.

    The layout of ``sksurv.util.Surv.from_arrays``: two fields, the first
    a boolean event indicator and the second the event or censoring time.
    """
    array = np.asarray(y)
    names = list(array.dtype.names or ())
    if (
        len(names) != 2
        or array.dtype[names[0]].kind != "b"
        or array.dtype[names[1]].kind not in "iuf"
    ):
        raise ValueError(
            "a structured survival array must have two fields, a boolean "
            f"event indicator then a numeric time; got fields {tuple(names)}"
        )
    events = np.ascontiguousarray(array[names[0]], dtype=np.bool_)
    times = np.ascontiguousarray(array[names[1]], dtype=np.float64)
    return Response("aft", "right", times=times, events=events)


def _interval_response(bounds: npt.NDArray[Any]) -> Response:
    """Parse an `(n, 2)` array of lower and upper bounds.

    An infinite bound is one-sided censoring and an equal pair an exact
    value, as the core reads them.
    """
    array = np.ascontiguousarray(bounds, dtype=np.float64)
    return Response(
        "interval_censored",
        "interval",
        lower=np.ascontiguousarray(array[:, 0]),
        upper=np.ascontiguousarray(array[:, 1]),
    )


def _as_response(y: Any) -> Response:
    """Parse `y` into the shape the core takes.

    Parameters
    ----------
    y : array_like
        The response as the caller gave it.

    Returns
    -------
    Response
        The parsed response.

    Raises
    ------
    ValueError
        For a shape no family takes, or a missing value.
    """
    categorical = _categorical(y)
    if categorical is not None:
        return _categorical_response(categorical)
    if getattr(getattr(y, "dtype", None), "names", None):
        return _survival_response(y)
    array = np.asarray(y)
    if array.dtype.kind == "b":
        return Response(
            "plain", "binary", y=np.ascontiguousarray(array, dtype=np.float64)
        )
    if array.ndim == 2 and array.shape[1] == 2:
        return _interval_response(array)
    return Response("plain", "numeric", y=_as_numeric(array))


def _family_name(outcome: Outcome) -> str:
    """Return the stored name of `outcome`'s family."""
    return next(iter(outcome._core()))


def _model_family(model: str) -> str:
    """Return the family of a fitted model's name."""
    return "gaussian" if model == "heteroscedastic" else model


def _check_family(family: str, response: Response) -> None:
    """Refuse a family the response does not fit.

    Parameters
    ----------
    family : str
        The stored name of the named family, or of a fitted model.
    response : Response
        The parsed response.

    Raises
    ------
    ValueError
        Naming the shape, the family it selects and the family named.
    """
    family = _model_family(family)
    if response.shape in FAMILY_SHAPES.get(family, ("numeric",)):
        return
    raise ValueError(
        f"The response is {SHAPE_LABEL[response.shape]}, which selects the "
        f"{response.family} family, but outcome names the {family} family."
    )


def _family_of(response: Response) -> Outcome:
    """Return the family the response selects, at its defaults."""
    family = response.family
    if family == "probit":
        return Probit()
    if family == "ordinal":
        return Ordinal(categories=len(response.categories or ()))
    if family == "aft":
        return Aft()
    if family == "interval_censored":
        return IntervalCensored()
    return Gaussian()


def _resolve_outcome(outcome: Outcome | None, response: Response) -> Outcome:
    """Return the outcome family resolved against the response.

    `None` takes the family the response selects; a named family is
    checked against the response and never coerced. The ordinal family
    takes its category count from an ordered `Categorical` where the
    constructor left it unset.

    Parameters
    ----------
    outcome : Outcome or None
        The family as configured.
    response : Response
        The parsed response.

    Returns
    -------
    Outcome
        A named family.

    Raises
    ------
    ValueError
        For a family the response does not fit, an ordinal category count
        disagreeing with the categories, or an ordinal family over integer
        codes with no category count.
    """
    if outcome is None:
        return _family_of(response)
    _check_family(_family_name(outcome), response)
    if not isinstance(outcome, Ordinal):
        return outcome
    if response.categories is None:
        if outcome.categories is None:
            raise ValueError(
                "ordinal() needs categories over integer codes; "
                "an ordered Categorical response carries its own"
            )
        return outcome
    count = len(response.categories)
    if outcome.categories is None:
        return Ordinal(
            categories=count, offset=outcome.offset, cutpoint_sd=outcome.cutpoint_sd
        )
    if outcome.categories != count:
        raise ValueError(
            f"outcome names {outcome.categories} categories but the response "
            f"has {count}"
        )
    return outcome


def _fit(
    config: str,
    design: Float,
    response: Response,
    seed: int,
    n_chains: int,
    n_threads: int,
) -> _native.Fitted:
    """Fit through the core entry point the response reaches."""
    if response.kind == "aft":
        return _native.fit_aft(
            config, design, response.times, response.events, seed, n_chains, n_threads
        )
    if response.kind == "interval_censored":
        return _native.fit_interval_censored(
            config, design, response.lower, response.upper, seed, n_chains, n_threads
        )
    return _native.fit(config, design, response.y, seed, n_chains, n_threads)


def _new_sampler(
    config: str, design: Float, response: Response, seed: int
) -> _native.Sampler:
    """Construct the core's sampler over a parsed response."""
    if response.kind == "aft":
        return _native.Sampler.aft(
            config, design, response.times, response.events, seed
        )
    if response.kind == "interval_censored":
        return _native.Sampler.interval_censored(
            config, design, response.lower, response.upper, seed
        )
    return _native.Sampler(config, design, response.y, seed)


def _set_response(sampler: _native.Sampler, response: Response) -> None:
    """Replace a sampler's response with a parsed one."""
    if response.kind == "aft":
        sampler.set_aft_response(response.times, response.events)
    elif response.kind == "interval_censored":
        sampler.set_interval_censored_response(response.lower, response.upper)
    else:
        sampler.set_response(response.y)


def _log_likelihood(fitted: _native.Fitted, design: Float, response: Response) -> Float:
    """Return the pointwise log-likelihood of a parsed response, draw-major."""
    if response.kind == "aft":
        return fitted.log_likelihood_survival(design, response.times, response.events)
    if response.kind == "interval_censored":
        return fitted.log_likelihood_interval_censored(
            design, response.lower, response.upper
        )
    return fitted.log_likelihood(design, response.y)


def _sampler_family(sampler: _native.Sampler) -> str:
    """Return the stored family name of a live sampler's configuration."""
    outcome: dict[str, Any] = json.loads(sampler.config)["outcome"]
    return next(iter(outcome))
